src/NHibernate/Dialect/TypeNames.cs

```diff
@@ -72,44 +72,65 @@ public string Get(DbType typecode)
 		/// <param name="scale">the SQL scale </param>
 		/// <param name="precision">the SQL precision </param>
 		/// <returns>
-		/// The associated name with smallest capacity >= size if available and the
-		/// default type name otherwise
+		/// The associated name with smallest capacity >= size (or scale for date time types) if available,
+		/// otherwise the default type name.
 		/// </returns>
 		public string Get(DbType typecode, int size, int precision, int scale)
 		{
-			SortedList<int, string> map;
-			weighted.TryGetValue(typecode, out map);
+			weighted.TryGetValue(typecode, out var map);
 			if (map != null && map.Count > 0)
 			{
-				foreach (KeyValuePair<int, string> entry in map)
+				var requiredCapacity = IsScaleType(typecode) ? scale : size;
```

Comment:
Now handling types for which capacity means scale, not size.
Something similar could be done for precision, but it would requires to fix the dialect current decimal registrations, which do not appear to transmit valid capacity for `decimal`. Moreover, the default for decimal would not suit when a "too big" capacity is asked, because this default has usually not the maximal precision. It would then be required to have some additional logic here for yielding the maximal actual precision here. But it would have to account scale too, and that would be quite messy. (Should it reduces the scale as much as precision is reduced? Or just trims it to max too? Or something in between?)